### PR TITLE
FIX: refactor dash scaling

### PR DESCRIPTION
**Pull request by tacaswell**
This start to address #6592 

The collection code still is broken.

### CODE REVIEW

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -24,12 +24,11 @@
 import matplotlib.transforms as transforms
 import matplotlib.artist as artist
 from matplotlib.artist import allow_rasterization
-import matplotlib.backend_bases as backend_bases
 import matplotlib.path as mpath
 from matplotlib import _path
 import matplotlib.mlab as mlab
 import matplotlib.lines as mlines
-
+import math
```
Comment (QuLogic) [inline lib/matplotlib/collections.py]: Should go at the beginning, really. And not remove the second blank line.

**Comment by tacaswell** [comment]
Also close #6588 

Supersedes / closes #6590

**Comment by tacaswell** [comment]
Closes #6693

so

from @afvincent  in the issue

```python
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.ticker import NullLocator


def plot_test(ax, ls=None, lw=None):
    _plot_opts = dict(facecolor='LightGrey', edgecolor='Black')
    ax.set_title("ls:'{ls}' & lw: {lw}".format(ls=ls, lw=lw))
    ax.fill_between([1, 2, 3], [4, 5, 6], 5, linestyle=ls, lw=lw, **_plot_opts)
    ax.xaxis.set_major_locator(NullLocator())
    ax.yaxis.set_major_locator(NullLocator())
    return ax




fig, ((ax0, ax1), (ax2, ax3)) = plt.subplots(nrows=2, ncols=2,
                                             num=mpl.__version__)

ax0 = plot_test(ax0, ls='dashed', lw=2)
ax1 = plot_test(ax1, ls='dashed', lw=10)
ax2 = plot_test(ax2, ls='dashed', lw=None)
ax3 = plot_test(ax3, ls='solid', lw=0)
```

and
so

```python
import matplotlib.pyplot as plt
plt.fill_between([1,2,3], [4,5,6], 5, linestyle="None")
```

**Comment by tacaswell** [comment]
Also closes #5430

File: lib/matplotlib/collections.py [hunk 2]
```diff
@@ -534,29 +544,40 @@ def set_linestyle(self, ls):
...
+
         except ValueError:
-            raise ValueError('Do not know how to convert %s to dashes' % ls)
-        self._linestyles = dashes
-        self.stale = True
+            raise ValueError(
+                'Do not know how to convert {!r} to dashes'.format(ls))
+
+        self._us_linestyles = dashes
+        self._linewidths
```
Comment (afvincent) [inline lib/matplotlib/collections.py]: I guess the line just above is a typo, isn't it?

Comment (tacaswell) [inline lib/matplotlib/collections.py]: Fixed.  Left over from me spell checking variables :sheep:

**Comment by mdboom** [comment]
I think the dash scaling needs to happen from both set_linestyle and set_linewidth.  Otherwise, this looks good.

File: lib/matplotlib/lines.py [hunk 3]
```diff
@@ -36,24 +36,47 @@
...
         dashes = tuple(rcParams['lines.{}_pattern'.format(style)])
+    #
     elif isinstance(style, tuple):
         offset, dashes = style
     else:
         raise ValueError('Unrecognized linestyle: %s' % str(style))
 
+    # normalize offset to be positive and shorter than the dash cycle
+    if dashes is not None and offset is not None:
+        offset %= sum(dashes)
```
Comment (afvincent) [inline lib/matplotlib/lines.py]: Don't know if one are supposed to be “bad user-proof”, but if `dashes` is (0, 0), won't it raise a `ZeroDivisionError`? (I know it's kind of a corner case though)

**Comment by mdboom** [comment]
Sorry -- I just missed it.  The context in the diff wasn't so clear.  LGTM.

File: lib/matplotlib/collections.py [hunk 4]
```diff
@@ -117,7 +122,10 @@ def __init__(self,
         """
         artist.Artist.__init__(self)
         cm.ScalarMappable.__init__(self, norm, cmap)
-
+        self._us_linestyles = [(None, None)]
```
Comment (afvincent) [inline lib/matplotlib/collections.py]: FWIW, it was not obvious to me that “_us” prefix stood for “unscaled” (at least until I read ~ l. 390 in `lines.py`). Adding a small comment might be useful for future readers.

**Comment by afvincent** [comment]
@tacaswell : Would it be possible to write a small docstring for the static method `_bcast_lwls`? (At least something that tells it broadcast linewidths and linestyles.)

**Comment by tacaswell** [comment]
@afvincent Thanks!

**Comment by afvincent** [comment]
You're welcome (but you're the one who did the hard work).

For my knowledge, about ll.606-609 in `collections.py`: 
```python
# pass in None as first arg as collection do not
# support offset in linestyles (yet)
dashes = [mlines._scale_dashes(o, d, lw)
          for (o, d), lw in zip(dashes, linewidths)]
```
where is `None` passed in as an argument? If `dashes` is a correct dash specification `(offset, (dash pattern tuple))`, I don't see any `None` on the horizon then. (But maybe I misunderstood the comment.)

**Comment by tacaswell** [comment]
That is an old comment (it was a windy road to sort out how to make this work) that is no longer true.

**Pull request merged**